contracts/rewards/src/contract/execute.rs

```diff
@@ -77,113 +77,106 @@ where
-            cur_epoch.epoch_num.saturating_sub(2),
+            (from + epoch_process_limit).saturating_sub(1), // for process limit =1 "from" and "to" must be equal
+            cur_epoch.epoch_num.saturating_sub(EPOCH_PAYOUT_DELAY),
         );
-        if to < from || cur_epoch.epoch_num < 2 {
+
+        if to < from || cur_epoch.epoch_num < EPOCH_PAYOUT_DELAY {
             return Err(ContractError::NoRewardsToDistribute.into());
         }
-        let rewards =
-            self.process_rewards_for_epochs(contract.clone(), from, to, HashMap::new())?;
-        self.store.save_rewards_watermark(contract, to)?;
+
+        let rewards = self.process_rewards_for_epochs(contract_addr.clone(), from, to)?;
+        self.store.save_rewards_watermark(contract_addr, to)?;
         Ok(rewards)
     }
 
     fn process_rewards_for_epochs(
         &mut self,
-        contract: Addr,
+        contract_addr: Addr,
         from: u64,
         to: u64,
-        rewards: HashMap<Addr, Uint128>,
-    ) -> Result<HashMap<Addr, Uint128>, ContractError> {
-        if from > to {
-            return Ok(rewards);
-        }
-        let new_rewards = self.process_rewards_for_epoch(contract.clone(), from)?;
-        self.process_rewards_for_epochs(contract, from + 1, to, merge_rewards(rewards, new_rewards))
-    }
-
-    pub fn process_rewards_for_epoch(
-        &mut self,
-        contract: Addr,
-        epoch_num: u64,
     ) -> Result<HashMap<Addr, Uint128>, ContractError> {
         self.store
-            .load_epoch_tally(contract, epoch_num)?
-            .map_or(Ok(HashMap::new()), |tally| self.process_epoch_tally(tally))
+            .load_rewards_pool(contract_addr.clone())?
+            .sub_reward(
+                self.cumulate_rewards(&contract_addr, from, to)
+                    .values()
+                    .sum(),
+            )?
+            .then(|pool| self.store.save_rewards_pool(&pool))?;
+
+        Ok(self.cumulate_rewards(&contract_addr, from, to))
     }
 
-    fn process_epoch_tally(
+    fn cumulate_rewards(
         &mut self,
-        tally: EpochTally,
-    ) -> Result<HashMap<Addr, Uint128>, ContractError> {
-        let workers_to_reward = tally.workers_to_reward();
-
-        let mut pool = self
-            .store
-            .load_rewards_pool(tally.contract.clone())?
-            .unwrap_or(RewardsPool {
-                contract: tally.contract,
-                balance: Uint128::zero(),
-            });
-
-        let rewards_per_worker =
-            rewards_per_worker(&workers_to_reward, tally.params.rewards_per_epoch)?;
-
-        pool.distribute_rewards(workers_to_reward.len() as u64, rewards_per_worker)?;
-        self.store.save_rewards_pool(&pool)?;
-
-        Ok(workers_to_reward
-            .into_iter()
-            .map(|worker| (worker, rewards_per_worker))
-            .collect())
+        contract_addr: &Addr,
+        from: u64,
+        to: u64,
+    ) -> HashMap<Addr, Uint128> {
+        (from..=to)
```

Comment:
I am honestly not convinced that this is cleaner than the recursion.

Reply:
Yeah I'm not sure that this cleaner. The recursive version seems a little easier to read and understand in my opinion, and is more concise:
```
        if from > to {
            return Ok(rewards);
        }
        let new_rewards = self.process_rewards_for_epoch(contract.clone(), from)?;
        self.process_rewards_for_epochs(contract, from + 1, to, merge_rewards(rewards, new_rewards))
```

Reply:
I'll give it another try, my issue with the recursion was that it went three layers deep

Reply:
to me, recursion is always the cleanest when the call stack has 3 layers lol.